### PR TITLE
#161290898 Signup Verification Email

### PR DESCRIPTION
**Pull request by ajimae**
#### What does this PR do?
 Enable users to receive a welcome message and an email verification link on successful registration

#### Description of Task to be completed?

- Modify migration and user model files to accommodate emailVerificationHash column
- Write functions that will enable users receive a verification link on successful registration
- Add a middleware script to check email/account verification status
- Write test to validate functionality when user is verified or unverified
- Refactor email library from sendGrid to nodemailer
- Refactor email template to comply with the team accepted format
- Write a function to update the database on successful email verification
- Write tests to validate email/account verification functionality

#### How should this be manually tested?

- Clone the repo
- Run `npm install` to install dependencies and devDependencies
- Run Postma, using the `POST` verb visit the address `localhost:3000/api/v1/users/register` to register an account
- If step 3 above is successful login to the provided email address and copy the link linked to the button and paste in Postman, using the `GET` verb click on send

#### Any background context you want to provide?

- A valid email and password may be provided in the env variables to be able to use the emailing services.

#### What are the relevant pivotal tracker stories?

[#161290898](https://www.pivotaltracker.com/story/show/161290898)

#### Screenshots

screen shot 2018-11-19 at 1 52 57 am

### CODE REVIEW

File: server/src/controllers/UserController.js [hunk 1]
```diff
@@ -58,10 +61,17 @@ class UserController {
...
+            to: email,
+            subject: 'Welcome to Authors Haven',
+            html: html(username, req.headers.host, UpdateWorker.hashGenerator(email))
+          };
+          Mailer.sendVerificationEmail(mailOption);
           return res.header('x-token', token).status(200).json({
             user: {
               success: true,
-              message: 'registration successful',
+              message: 'Registration successful',
```
Comment (ovieokeh) [inline server/src/controllers/UserController.js:78]: I thought we were using all lowercase for the response messages?

File: server/src/db/migrations/a-create-user.js [hunk 2]
```diff
@@ -1,3 +1,21 @@
+const variableTest = (Sequelize) => {
+  let variable;
+  if (process.env.NODE_ENV === 'test') {
```
Comment (ovieokeh) [inline server/src/db/migrations/a-create-user.js:3]: Perfect idea Chukwuemeka!!

**Review by ovieokeh — APPROVED** [review]
I've gone through all the files changed. Awesome work!

File: server/src/db/migrations/a-create-user.js (hunk 2 above)
Comment (oluseyianani) [reply server/src/db/migrations/a-create-user.js:3]: This is amazing @ajimae. Great Job!

**Review by oluseyianani — APPROVED** [review]
Changes look good. Nodemailer works better than sendgrid API after much email blasts.

**Comment by desaawa** [comment]
Please resolve conflicts and failing tests

**Review by hope-uwa — CHANGES_REQUESTED** [review]
@ajimae Great job on integrating nodemailer for sending emails. I noticed you disabled Airbnb eslint console.log rule on some lines in the sendMail.js file. I believe its best practice that console shouldn’t be allowed on production code. The nodemailer errors displayed during the tests makes the test buggy and hard to read. I think you can capture the errors some other way, for example by logging it to a text file. Thanks.

File: server/src/helpers/articleRatings.js [hunk 3]
```diff
@@ -66,11 +66,13 @@ class Rating {
       userId,
       articleId,
       ratings: rateValue,
-    }).then(data => res.status(201).json({
-      data,
-      success: true,
-      message: 'You successfully rated this article'
-    }));
+    }).then((data) => {
```
[the remark this answers is not in the document]
Comment (ovieokeh) [reply server/src/helpers/articleRatings.js]: To solve this issue, just remove the `{ return` part of `.then` segment. Plus the merge conflicts and other Hound CI errors should also be rectified.

**Review by hope-uwa — APPROVED** [review]

**Review by desaawa — APPROVED** [review]

**Pull request merged**